### DISCUSSION TITLE
[dynamic-Api-Routes]从 api 动态加载路由。帮助开发插件。

### DISCUSSION THREAD

**Discussion by xXAvoraXx** [umijs/umi#11770] [Ideas]
https://github.com/xXAvoraXx/plugin-dynamicApiRoutes

大家好、

**该插件的目的是**
1.登录后，用户应向配置文件 "dynamicApiRoutes.requestUrl "中定义的 URL 发送请求，并从 api 获取路由定义。
2.路由应动态创建，而非静态创建。

**为什么要从 api 获取路由？**
1.之所以要从 api 中定义路由，是因为我们想动态控制路由，例如，我们想定义一条新路由到数据库，并将该路由与我们想要的组件相匹配。
2.我们使用基于 ReBAC 的授权，而不是基于 RBAC 的授权，因此每个用户的路由和菜单定义可能不同。因此，用户登录后，"授权层 "会立即通知我们用户拥有哪些路由和菜单。

请提出建设性的批评意见，静态路由并不能满足我们的要求。我们不希望用户从 BrowserRouter 中学习到所有路由。

https://github.com/ant-design/ant-design-pro/discussions/10898

**Comment by xXAvoraXx** [comment]
检查用户是否登录的插件开发路线图。如果您有不同的想法，请补充进来。

1. 检查 cookie 中名为 "access_token "的属性是否存在以及令牌是否有效。
2. 检查 localStorage 中名为 "access_token "的属性是否存在以及令牌是否有效。
3. 可以检查 initialState 中的状态值。
4. 应在配置文件中指定要检查的实体名称 "access_token"。
5. 如果用户已登录，则应从 api 调用动态路由。

**Reply by xXAvoraXx** [reply]
调用应用程序时，patchRoutes 函数会被调用一次，因此在用户登录系统后，必须借助钩子刷新页面。

**Reply by xXAvoraXx** [reply]
代码中的 PageLoading 组件应该是可定制的，它应该能够引用配置文件。
https://github.com/xXAvoraXx/plugin-dynamicApiRoutes/blob/master/src/utils/getLazyLoadableContent.ts#L3

路由图标应从 "public "文件夹中定义或作为链接。
https://github.com/xXAvoraXx/plugin-dynamicApiRoutes/blob/master/src/utils/getRuntimeContent.ts#L4

**Comment by xXAvoraXx** [comment]
https://github.com/xXAvoraXx/plugin-dynamicApiRoutes/discussions/1

**Comment by whiteshader** [comment]
从api加载动态路由请参考以下项目。
https://github.com/whiteshader/React-Admin-Cloud

**Reply by xXAvoraXx** [reply]
能否举例说明如何从 api 加载动态路由？您发送的项目已经是一个 pro.ant.design 项目。React-Admin-Cloud 如何从 api 动态加载路由？

**Reply by whiteshader** [reply]
请看app.tsx

**Reply by xXAvoraXx** [reply]
@whiteshader 您是否会考虑将其作为附加功能来编写？

**Reply by whiteshader** [reply]
umi这块不够自由又复杂，打算将umi砍了

**Reply by xXAvoraXx** [reply]
您愿意为以下开发流程做出贡献吗？
https://github.com/xXAvoraXx/plugin-dynamicApiRoutes

**Reply by xXAvoraXx** [reply]
@whiteshader  我应用了您的示例，但菜单的语言并没有改变。

**Comment by whiteshader** [comment]
从api加载动态路由请参考以下项目。
https://github.com/whiteshader/React-Admin-Cloud
